Description: Make it safer to run loadkeys with foreground console in raw mode
 If not invoked with an explicit console parameter, try to find any console
 not in raw mode.  This makes it safer to run loadkeys if the foreground
 console might be in raw mode, for example if Plymouth or X might be
 running.
Author: Colin Watson <cjwatson@ubuntu.com>
Author: Steve Langasek <steve.langasek@ubuntu.com>
Forwarded: no
Last-Update: 2016-09-20

Index: kbd-1.15.5/src/getfd.c
===================================================================
--- kbd-1.15.5.orig/src/getfd.c
+++ kbd-1.15.5/src/getfd.c
@@ -86,3 +86,56 @@
 	/* total failure */
 	exit(1);
 }
+
+static int
+kbmode(int fd) {
+	int mode;
+
+	if (ioctl(fd, KDGKBMODE, &mode) < 0)
+		return -1;
+	else
+		return mode;
+}
+
+/* Get a file descriptor suitable for setting the keymap.  Anything not in
+ * raw mode will do - it doesn't have to be the foreground console - but we
+ * prefer Unicode if we can get it.
+ */
+int
+getfd_keymap(void) {
+	int fd, bestfd = -1;
+	char ttyname[sizeof("/dev/ttyNN")] = "/dev/tty";
+	int i;
+
+#define CHECK_FD_KEYMAP do { \
+	if (is_a_console(fd)) { \
+		int mode = kbmode(fd); \
+		if (mode == K_UNICODE) { \
+			if (bestfd != -1) \
+				close(bestfd); \
+			return fd; \
+		} else if (mode == K_XLATE && bestfd == -1) \
+			bestfd = fd; \
+		else \
+			close(fd); \
+	} else \
+		close(fd); \
+} while (0)
+
+	for (i = 0; conspath[i]; i++) {
+		if ((fd = open_a_console(conspath[i])) >= 0)
+			CHECK_FD_KEYMAP;
+	}
+
+	for (i = 1; i <= 12; ++i) {
+		snprintf(ttyname + sizeof("/dev/tty") - 1, 3, "%d", i);
+		if ((fd = open_a_console(ttyname)) >= 0)
+			CHECK_FD_KEYMAP;
+	}
+
+	fprintf(stderr,
+		_("Couldn't get a file descriptor referring to the console\n"));
+
+	/* total failure */
+	exit(1);
+}
Index: kbd-1.15.5/src/getfd.h
===================================================================
--- kbd-1.15.5.orig/src/getfd.h
+++ kbd-1.15.5/src/getfd.h
@@ -2,5 +2,6 @@
 #define _GETFD_H
 
 extern int getfd(const char *fnam);
+extern int getfd_keymap(void);
 
 #endif /* _GETFD_H */
Index: kbd-1.15.5/src/loadkeys.y
===================================================================
--- kbd-1.15.5.orig/src/loadkeys.y
+++ kbd-1.15.5/src/loadkeys.y
@@ -1149,7 +1149,10 @@
 	}
 
 	/* get console */
-	fd = getfd(console);
+	if (console)
+		fd = getfd(console);
+	else
+		fd = getfd_keymap();
 
 	if (!optm && !optb) {
 		/* check whether the keyboard is in Unicode mode */
